File: 870-magic-squares-in-grid/magic-squares-in-grid.js
var numMagicSquaresInside = function(grid) {
    let rows = grid.length
    let columns = grid[0].length
    let count = 0

    function isMagic(row, col) {
        let lookup = new Set()

        for (let i = row; i < row + 3; i++) {
            for (let j = col; j < col + 3; j++) {
                let cell = grid[i][j]
                if (lookup.has(cell) || cell < 1 || cell > 9) return false
                lookup.add(cell)
            }
        }
        for (let i = row; i < row + 3; i++) {
            let row_sum = 0
            for (let j = col; j < col + 3; j++) {
                row_sum += grid[i][j]
            }
            if (row_sum != 15) return false
        }

        for (let j = col; j < col + 3; j++) {
            let col_sum = 0
            for (let i = row; i < row + 3; i++) {
                col_sum += grid[i][j]
            }
            if (col_sum != 15) return false
        }

        let primary_diag = 0
        primary_diag += grid[row][col] + grid[row + 1][col + 1] + grid[row + 2][col + 2]

        if (primary_diag != 15) return false


        let secondary_diag = 0
        secondary_diag += grid[row][col + 2] + grid[row + 1][col + 1] + grid[row + 2][col]
        if (secondary_diag != 15) return false

        return true
    }

    for (let i = 0; i < rows - 2; i++) {
        for (let j = 0; j < columns - 2; j++) {
            if (isMagic(i, j)) {
                count += 1
            }
        }
    }

    return count
};

File: 870-magic-squares-in-grid/magic-squares-in-grid.py
class Solution:
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        columns = len(grid[0])
        result = 0

        def isMagic(r, c):
            # validate cells in 3 * 3 grid are unique
            lookup = set()

            for i in range(r, r + 3):
                for j in range(c, c + 3):
                    cell = grid[i][j]
                    
                    if (cell in lookup or not (1 <= cell <= 9)):
                        return 0
                    lookup.add(cell)


            # validate rows
            for i in range(r, r + 3):
                row_sum = sum(grid[i][c : c + 3])

                if (row_sum != 15):
                    return 0                

            # validate columns
            for j in range(c, c + 3):
                column_sum = 0
                for i in range(r, r + 3):
                    column_sum += grid[i][j]

                if (column_sum != 15):
                    return 0

            # validate diagonals
            if (grid[r][c] + grid[r + 1][c + 1] + grid[r + 2][c + 2] != 15) or (grid[r][c + 2] + grid[r + 1][c + 1] + grid[r + 2][c] != 15):
                return 0

            return 1


        for i in range(rows - 2):
            for j in range(columns - 2):
                result += isMagic(i, j)


        return result